lib/util/source-code.js

```diff
@@ -191,25 +202,61 @@ SourceCode.prototype = {
      * @public
      */
     getComments(node) {
+        if (this._commentStore.has(node)) {
+            return this._commentStore.get(node);
+        }
 
-        let leadingComments = node.leadingComments || [];
-        const trailingComments = node.trailingComments || [];
+        const comments = {
+            leading: [],
+            trailing: []
+        };
 
         /*
-         * espree adds a "comments" array on Program nodes rather than
-         * leadingComments/trailingComments. Comments are only left in the
-         * Program node comments array if there is no executable code.
+         * Espree only leaves comments in the Program node comments
+         * array if there is no executable code.
          */
         if (node.type === "Program") {
             if (node.body.length === 0) {
-                leadingComments = node.comments;
+                comments.leading = node.comments;
+            }
+        } else if ((node.type === "BlockStatement" || node.type === "ClassBody") && node.body.length === 0 ||
+            node.type === "ObjectExpression" && node.properties.length === 0 ||
+            node.type === "ArrayExpression" && node.elements.length === 0 ||
+            node.type === "SwitchStatement" && node.cases.length === 0
+        ) {
+            const tokens = this._getTokens(node);
+
+            comments.trailing = tokens.filter(astUtils.isComment);
```

Comment:
Adding comments as leading comments to an otherwise empty "Program" node but adding them as trailing comments to other empty nodes is a holdover from Espree. Is this something we'd like to keep? I didn't change it in my first pass at this to keep the behavior as similar as possible, but this seems highly like to cause confusion in the future...